epm.js

```diff
@@ -75,31 +75,37 @@ EPM.prototype.resolve = function(import_path, imported_from, callback) {
   var separator = import_path.indexOf("/");
   var package_name = import_path.substring(0, separator);
   var internal_path = import_path.substring(separator + 1);
-  var install_directory = path.join(this.working_directory, "installed_contracts");
-
-  var resolved_contents = undefined;
-  var resolved_path = undefined;
-
-  detectSeries([
-    path.join(install_directory, import_path),
-    path.join(install_directory, package_name, "contracts", internal_path)
-  ], function(file_path, finished) {
-    fs.readFile(file_path, {encoding: "utf8"}, function(err, body) {
-      if (err) return finished(null, false);
-
-      resolved_contents = body;
-      resolved_path = file_path;
-
-      finished(null, true);
-    });
-  }, function(err, existing_path) {
-    // If there's an error, that means we can't read the source even if
-    // it exists. Treat it as if it doesn't by ignoring any errors.
-    // Perhaps we can do something better here in the future.
+  var installDir = this.working_directory;
+
+  // If nothing's found, body returns `undefined`
+  var body;
+
+  while(true){
+    var file_path = path.join(installDir, "installed_contracts", import_path);
+
+    try {
+      body = fs.readFileSync(file_path, {encoding: "utf8"});
+      break;
+    }
+    catch(err){}
+
+    file_path = path.join(installDir, "installed_contracts", package_name, "contracts", internal_path)
+
+    try {
+      body = fs.readFileSync(file_path, {encoding: "utf8"});
+      break;
+    }
+    catch(err){}
+
+    // Recurse outwards until impossible
+    var oldInstallDir = installDir;
+    installDir = path.join(installDir, '..');
+    if (installDir === oldInstallDir) {
+      break;
+    }
+  }
```

Comment:
Can we abstract the `while(true)` block so we don't have to define it in two places?

Reply:
I'm open to it - if they both only checked one path format it would be perfect. But epm checks two which starts to make the abstraction kind of complicated for what this is... ok I'll think about it for bit, maybe something nice is possible.
```
path.join(installDir, "installed_contracts", import_path);
path.join(installDir, "installed_contracts", package_name, "contracts", internal_path)
```

Reply:
Ah, maybe something like this?
```javascript
/**
 * @param importPathFunc - function or array of functions converting importPath -> filePath
 */
function resolveUp(importPath, installDir, importPathFunc, callback) {
  var importPathFuncs = Array.isArray(importPathFunc)
    ? importPathFunc
    : [ importPathFunc ];

  var filePath;
  var body;
  while (true) {
    for (let func of importPathFuncs) {
      filePath = func(importPath);

      try {
        body = fs.readFileSync(filePath, {encoding: "utf8"});
        return body;
      } catch(err) {}
    }

    // Recurse outwards until impossible
    var oldInstallDir = installDir;
    installDir = path.join(installDir, '..');
    if (installDir === oldInstallDir) {
      break;
    }
  }
}
```

Reply:
Right - it's possible - do you think this is better? I'm on the fence about it - you have to disambiguate between the cases and add an additional loop, then locate this in a place where it's shared - also define the `funcs` and make the callbacks. There's like a slow decay of legibility for something that's very simple despite the duplication. 

I wonder if people are going to start hoisting their build artifacts though, because then this is **definitely** going in.

Reply:
```
build/
node_modules/
packages/
truffle.js
```

Reply:
I am also on the fence about it, but I think the deduplication is slightly preferable in terms of codifying this recursive resolution because it makes it explicit.

My motivation is that I don't like the separation of NPM / "EPM"... I don't like that one can break but not the other; it'd be nice if the import stuff was more unified to prevent this.

Reply:
(Normally the rule of thumb is 3 copypastas, so that's my advocating to make an exception there)